regcomp.c

```diff
@@ -24850,6 +24878,383 @@ S_parse_uniprop_string(pTHX_
+                                         this_name_start - all_names_start);
+
+            /* If we didn't find a \n, it means it matched somewhere in the
+             * initial '00000' in the string, so isn't a real match */
+            if (cp_end == NULL) {
+                continue;
+            }
+
+            this_name_start = cp_end + 1;   /* The name starts just after */
+            cp_end--;                       /* the \n, and the code point */
+                                            /* ends just before it */
+
+            /* All code points are 5 digits long */
+            cp_start = cp_end - 4;
+
+            /* Except for the first line in the string, the sequence before the
+             * code point is \n\n.  If that isn't the case here, we didn't
+             * match the name of a character.  (We could have matched a named
+             * sequence, not currently handled */
+            if (      cp_start > all_names_start
+                && (*(cp_start - 1) != '\n' || *(cp_start - 2) != '\n'))
+            {
+                continue;
+            }
+
+            /* Calculate the code point from its 5 digits */
+            cp = (XDIGIT_VALUE(cp_start[0]) << 16)
+               + (XDIGIT_VALUE(cp_start[1]) << 12)
+               + (XDIGIT_VALUE(cp_start[2]) << 8)
+               + (XDIGIT_VALUE(cp_start[3]) << 4)
+               +  XDIGIT_VALUE(cp_start[4]);
+
+            /* We matched!  Add this to the list */
+            *prop_definition = add_cp_to_invlist(*prop_definition, cp);
+            found_matches = TRUE;
+        } /* End of loop through the non-algorithmic names string */
+    }
+
+    SvREFCNT_dec_NN(names_string);
+
+    /* There are also character names not in 'names_string'.  These are
+     * algorithmically generatable.  Try this pattern on each possible one.
+     * (khw originally planned to leave this out given the large number of
+     * matches attempted; but the speed turned out to be quite acceptable
+     *
+     * There are plenty of opportunities to optimize to skip many of the tests.
+     * beyond the rudimentary ones already here */
+
+    /* First see if the subpattern matches any of the algorithmic generatable
+     * Hangul syllable names.
+     *
+     * We know none of these syllable names will match if the input pattern
+     * requires more bytes than any syllable has, or if the input pattern only
+     * matches an empty name, or if the pattern has something it must match and
+     * one of the characters in that isn't in any Hangul syllable. */
+    if (    prog->minlen <= (SSize_t) syl_max_len
+        &&  prog->maxlen > 0
+        && ( ! must
+            || SvCUR(must) == 0
+            || strspn(SvPVX(must), "\n ABCDEGHIJKLMNOPRSTUWY") == SvCUR(must)))
+    {
+        /* These constants, names, values, and algorithm were taken from the
+         * Unicode standard, version 5.1, section 3.12, and should never
+         * change. */
```

Comment:
Maybe so, but it seems unnecessarily obfuscatory - LBase, VBase and TBase should not appear at all, and it should use int LCount = 19 instead of SCount and then:
```
  for (int L = 0; L < LCount; L++) {
    for (int V = 0; V < VCount; V++) {
      for (int T = 0; T < TCount; T++) {
        ...
```
.. which then leaves much more obvious the room for optimization by calling sv_catpv at each level. That and a comment explaining SBase would improve it quite a chunk.

Reply:
ok

Reply:
ok